File: src/main/scala/com/calclavia/graph/core/base/NodeBlockConnect.java
package com.calclavia.graph.core.base;

import com.calclavia.graph.api.node.Node;
import nova.core.block.Block;
import nova.core.component.ComponentProvider;
import nova.core.util.Direction;
import nova.core.util.exception.NovaException;
import nova.core.util.transform.vector.Vector3i;
import nova.core.world.World;

import java.util.Arrays;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * A node that connects to adjacent blocks.
 * @author Calclavia
 */
public abstract class NodeBlockConnect<N extends Node<?>> extends NodeConnect<N> {

	public final ComponentProvider provider;
	//The cached connection map
	protected Map<N, Direction> connectedMap;

	public NodeBlockConnect(ComponentProvider parent) {
		this.provider = parent;
	}

	@Override
	public Set<N> connections() {

		Map<Direction, Optional<Block>> adjacentBlocks = adjacentBlocks();

		Map<Direction, N> adjacentNodes = adjacentBlocks
			.entrySet()
			.stream()
			.filter(entry -> entry.getValue().isPresent())
			.filter(entry -> entry.getValue().get().getClass().isAssignableFrom(compareClass()))
			.collect(Collectors.toMap(Map.Entry::getKey, entry -> getNodeFromBlock(entry.getValue().get(), entry.getKey())));

		//Generates a map of connections and their directions
		connectedMap = adjacentNodes
			.entrySet()
			.stream()
			.filter(entry -> canConnect(entry.getValue(), entry.getKey()))
			.filter(entry -> ((NodeConnect) entry.getValue()).canConnect((N) this, entry.getKey().opposite()))
			.collect(Collectors.toMap(Map.Entry::getValue, Map.Entry::getKey));

		//Sets the connection mask based on the connections
		connectedMask = connectedMap.values().stream().map(Enum::ordinal).map(i -> 1 << i).reduce((a, b) -> a | b).orElseGet(() -> 0);

		return connectedMap.keySet();
	}

	protected Map<Direction, Optional<Block>> adjacentBlocks() {
		return Arrays.stream(Direction.DIRECTIONS)
			.collect(Collectors.toMap(Function.identity(), dir -> world().getBlock(position().add(dir.toVector()))));
	}

	protected N getNodeFromBlock(Block block, Direction from) {
		if (block instanceof ComponentProvider) {
			return (N) block.components()
				.stream()
				.filter(n -> compareClass().isAssignableFrom(n.getClass()))
				.findFirst()
				.orElse(null);
		}

		return null;
	}

	@Override
	public ComponentProvider provider() {
		return provider;
	}

	public World world() {
		if (provider instanceof Block) {
			return ((Block) provider).world();
		}
		throw new NovaException("NodeProvider type not supported.");
	}

	public Vector3i position() {
		if (provider instanceof Block) {
			return ((Block) provider).position();
		}
		throw new NovaException("NodeProvider type not supported.");
	}

	protected <N> Class<N> compareClass() {
		return (Class<N>) getClass();
	}
}

File: src/main/scala/com/calclavia/graph/core/base/NodeBlockConnect.scala
package com.calclavia.graph.core.base

import java.util.function.Supplier
import java.util.{Map => JMap, Optional, Set => JSet}

import com.calclavia.graph.api.Node
import nova.core.block.Block
import nova.core.util.Direction
import nova.core.util.transform.vector.Vector3i
import nova.core.world.World

import scala.collection.convert.wrapAll._

/**
 * A node that connects to adjacent blocks.
 * @author Calclavia
 */
trait NodeBlockConnect[N <: Node[N]] extends NodeConnect[N] {

	protected def block: Block

	//TODO: Expose to Java side
	protected var connectionFunction: () => JSet[N] = () => {
		val adjacentBlocks: Map[Direction, Optional[Block]] = this.adjacentBlocks
		val adjacentNodes: Map[Direction, N] =
			adjacentBlocks
				.filter { case (k, v) => v.isPresent && v.get.getClass().isAssignableFrom(compareClass) }
				.map { case (k, v) => (k, getNodeFromBlock(v.get(), k)) }

		val connectedMap = adjacentNodes
			.filter { case (k, v) => canConnect(v, k) }
			.filter { case (k, v) => v.asInstanceOf[NodeConnect[N]].canConnect(this.asInstanceOf[N], k.opposite) }
			.map(_.swap)

		connectedMask = connectedMap.values
			.map(_.ordinal)
			.map(i => 1 << i)
			.foldLeft(0)((a, b) => a | b)

		connectedMap.keySet
	}

	def setConnectionHandler(conFunction: () => JSet[N]): this.type = {
		connectionFunction = conFunction
		return this
	}

	def setConnections(f: Supplier[Set[N]]) {
		connectionFunction = () => f.get()
	}

	def setConnections(f: () => Set[N]) {
		connectionFunction = () => f.apply()
	}

	def connections: JSet[N] = connectionFunction()

	protected def adjacentBlocks: Map[Direction, Optional[Block]] = Direction.DIRECTIONS.map(dir => (dir, world.getBlock(position + dir.toVector))).toMap

	def world: World = block.world()

	def position: Vector3i = block.transform.position

	protected def getNodeFromBlock(block: Block, from: Direction): N = block.getOp(compareClass).orElse(null.asInstanceOf[N])

	protected def compareClass: Class[N] = getClass.asInstanceOf[Class[N]]
}